Source/JavaScriptCore/wasm/js/JSWebAssemblyStruct.cpp

```diff
@@ -0,0 +1,229 @@
+#include "JSWebAssemblyStruct.h"
+
+#if ENABLE(WEBASSEMBLY)
+
+#include "JSCInlines.h"
+#include "JSWebAssemblyInstance.h"
+#include "WasmFormat.h"
+#include "WasmModuleInformation.h"
+
+namespace JSC {
+
+constexpr size_t typeKindSizeInBytes(Wasm::TypeKind kind)
+{
+    using Wasm::TypeKind;
+
+    switch (kind) {
+    case TypeKind::I32:
+    case TypeKind::F32: {
+        return 4;
+    }
+    case TypeKind::I64:
+    case TypeKind::F64: {
+        return 8;
+    }
+
+    case TypeKind::Arrayref:
+    case TypeKind::Funcref:
+    case TypeKind::Externref:
+    case TypeKind::Ref:
+    case TypeKind::RefNull: {
+        return sizeof(WriteBarrierBase<Unknown>);
+    }
+    case TypeKind::Array:
+    case TypeKind::Func:
+    case TypeKind::Struct:
+    case TypeKind::Void:
+    case TypeKind::Rec:
+    case TypeKind::I31ref: {
+        break;
+    }
+    }
+
+    ASSERT_NOT_REACHED();
+    return 0;
+}
+
+const ClassInfo JSWebAssemblyStruct::s_info = { "WebAssembly.Struct"_s, &Base::s_info, nullptr, nullptr, CREATE_METHOD_TABLE(JSWebAssemblyStruct) };
+
+JSWebAssemblyStruct::JSWebAssemblyStruct(VM& vm, Structure* structure, Ref<Wasm::TypeDefinition>&& type)
+    : Base(vm, structure)
+    , m_type(WTFMove(type))
+    , m_payload(structPayloadSize(structType()))
+{
+    m_payload.fill(0);
+}
+
+JSWebAssemblyStruct* JSWebAssemblyStruct::tryCreate(JSWebAssemblyInstance* instance, uint32_t typeIndex)
+{
+    JSGlobalObject* globalObject = instance->globalObject();
+    Ref<Wasm::TypeDefinition> type = instance->instance().module().moduleInformation().typeSignatures[typeIndex];
+
+    VM& vm = globalObject->vm();
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
+
+    if (!globalObject->webAssemblyEnabled()) {
+        throwException(globalObject, throwScope, createEvalError(globalObject, globalObject->webAssemblyDisabledErrorMessage()));
+        return nullptr;
+    }
+
+    if (!Options::useWebAssemblyGC()) {
+        throwException(globalObject, throwScope, createEvalError(globalObject, "WebAssembly GC is not enabled."_s));
+        return nullptr;
+    }
+
+    auto* structValue = new (NotNull, allocateCell<JSWebAssemblyStruct>(vm)) JSWebAssemblyStruct(vm, globalObject->webAssemblyStructStructure(), WTFMove(type));
+    structValue->finishCreation(vm);
+    return structValue;
+}
+
+size_t JSWebAssemblyStruct::structPayloadSize(const Wasm::StructType* structType)
+{
+    size_t totalSize = 0u;
+    for (Wasm::StructFieldCount i = 0; i < structType->fieldCount(); ++i)
+        totalSize += typeKindSizeInBytes(structType->field(i).type.kind);
+    return WTF::roundUpToMultipleOf<sizeof(uint64_t)>(totalSize);
+}
+
+size_t JSWebAssemblyStruct::getFieldOffset(uint32_t fieldIndex, const Wasm::StructType* type)
+{
+    ASSERT(fieldIndex < type->fieldCount());
+
+    size_t fieldOffset = 0;
+    for (unsigned i = 0; i < fieldIndex; ++i)
+        fieldOffset += typeKindSizeInBytes(type->field(i).type.kind);
+    return fieldOffset;
+}
```

Comment:
Computing this every time accessing field is very costly. We should put this offset information to StructType.